packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,748 @@
+
+    // This factor controls the scaling from linear to logarithmic.
+    // scale = 0.0 -> fully linear scaling
+    // scale = 1.0 -> fully logarithmic scaling
+    let scale = 0.91; // Adjust this value between 0.0 and 1.0
+    for (let x = 0; x < targetSize; x++) {
+      // Linear interpolation between linear and log scaling
+      let linearIndex = (x / (targetSize - 1)) * (maxFreqIndex - 1);
+      let logScaledIndex =
+        logMinFreqIndex +
+        ((logMaxFreqIndex - logMinFreqIndex) * x) / (targetSize - 1);
+      let logIndex = Math.pow(10, logScaledIndex);
+
+      // Interpolating between linear and logarithmic scaling
+      let scaledIndex = (1.0 - scale) * linearIndex + scale * logIndex;
+
+      let index1 = Math.floor(scaledIndex);
+      let index2 = Math.ceil(scaledIndex);
+
+      if (index1 >= maxFreqIndex) {
+        index1 = maxFreqIndex - 1;
+      }
+      if (index2 >= maxFreqIndex) {
+        index2 = maxFreqIndex - 1;
+      }
+
+      if (index1 == index2) {
+        sample[x] = outSpectralData[index1];
+      } else {
+        let frac2 = scaledIndex - index1;
+        let frac1 = 1.0 - frac2;
+        sample[x] =
+          frac1 * outSpectralData[index1] + frac2 * outSpectralData[index2];
+      }
+    }
+
+    for (let x = 0; x < maxWidth; x++) {
+      // Based on research of looking at Winamp 5.666 and 2.63 executables
+
+      // if our bandwidth is "wide", chunk every 5 instances of the bars,
+      // add them together and display them
+      if (this._vis.bandwidth === "wide") {
+        chunk = chunk = x & 0xfffffffc;
+        uVar12 =
+          (sample[chunk + 3] +
+            sample[chunk + 2] +
+            sample[chunk + 1] +
+            sample[chunk]) /
+          4;
+        saData[x] = uVar12;
+      } else {
+        saData[x] = sample[x];
+      }
+
+      if (saData[x] >= maxHeight) {
+        saData[x] = maxHeight;
+      }
+
+      // prevents saPeaks going out of bounds when switching to windowshade mode
+      if (saPeaks[x] >= maxHeight * 256) {
+        saPeaks[x] = maxHeight * 256;
+      }
+
+      saFalloff[x] -= falloff / 16.0;
+      // Possible bar fall off values are
+      // 3, 6, 12, 16, 32
+      // Should there ever be some form of config options,
+      // these should be used
+      // 12 is the default of a fresh new Winamp installation
+
+      if (saFalloff[x] <= saData[x]) {
+        saFalloff[x] = saData[x];
+      }
+
+      if (saPeaks[x] <= Math.round(saFalloff[x] * 256)) {
+        saPeaks[x] = saFalloff[x] * 256;
+        saData2[x] = 3.0;
+      }
+
+      barPeak[x] = saPeaks[x] / 256;
+
+      saPeaks[x] -= Math.round(saData2[x]);
+      saData2[x] *= peakFalloff;
+      // Possible peak fall off values are
+      // 1.05f, 1.1f, 1.2f, 1.4f, 1.6f
+      // 1.1f is the default of a fresh new Winamp installation
+      if (saPeaks[x] <= 0) {
+        saPeaks[x] = 0;
+      }
+
+      if (smallVis) {
+        // SORRY NOTHING
```

Comment:
😆

Reply:
It's my mark! You know I was here when you see comments like these. 😄